### PR TITLE
[react-redux] Improved factory function types and deprecated `TypedUseSelectorHook`.

### PR DESCRIPTION
**Pull request by jrandolf-old**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [ ] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes:
See [React Redux | Custom Hooks](https://react-redux.js.org/api/hooks#custom-context) and the [source code](https://github.com/reduxjs/react-redux/blob/master/src/hooks) (each file contains a `createSomething` function which is typed in this package. The current typing for the `createEtc` functions reflect basic return and argument types, but **does not** reflect the relation between the argument and return type. This is especially important (in fact, without the relation, the type is useless) because the return type cannot be used, in general, when the argument is non-trivial (i.e. the argument is not {}). The depreciation of the `TypedUseSelectorHook` is because the new types will obsolete that type.
- [X] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [X] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/react-redux/index.d.ts
Comment (rbuckton) [inline types/react-redux/index.d.ts:576]:
```suggestion
 * @deprecated Use `createSelectorHook<State, Action>()`
```

**Review by rbuckton — CHANGES_REQUESTED** [review]

**Comment by andrewbranch** [comment]
This definitely needs to be reviewed by an owner before merging.

**Comment by jun-sheaf** [comment]
@kazuma1989 -- Mind reviewing this PR so we can merge it?

**Comment by jun-sheaf** [comment]
🔔 @tkqubo @kenzierocks @clayne11 @tansongyang @nicholasboll @mdibyo @kallikrein @val1984 @jrakotoharisoa @apapirovski @surgeboris @soerenbf @mrwolfz @dylanvann @Lazyuki @kazuma1989 — please [review this PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/46446/files). It's been up for quite a while. It would be good to have this merged by now.

**Comment by jun-sheaf** [comment]
@andrewbranch So it seems this PR even after two months has yet to be reviewed. It doesn't seem like any of the contributors are active. I have added tests to ensure the types are accurate previously. Could we get a maintainer to approve this request?

**Comment by andrewbranch** [comment]
> - [X] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>

@jun-sheaf you checked this box but didn’t do what it says 😞 In the absence of anyone who knows the package reviewing it, you’re going to have to give some kind of justification for this change.

**Comment by jun-sheaf** [comment]
@andrewbranch Thanks for getting back.

Sorry about that! I have edited the original comment and wrote in the necessary information.

I'll write it here as well
> See [React Redux | Custom Hooks](https://react-redux.js.org/api/hooks#custom-context) and the [source code](https://github.com/reduxjs/react-redux/blob/master/src/hooks) (each file contains a `createSomething` function which is typed in this package. The current typing for the `createEtc` functions reflect basic return and argument types, but **does not** reflect the relation between the argument and return type. This is especially important (in fact, without the relation, the type is useless) because the return type cannot be used, in general, when the argument is non-trivial (i.e. the argument is not {}). The depreciation of the `TypedUseSelectorHook` is because the new typings will obsolete that type.

**Review by andrewbranch — APPROVED** [review]
Thanks for the explanation; makes sense to me 👍

**Comment by andrewbranch** [comment]
@jun-sheaf since none of the existing owners showed up to review this, would you consider adding yourself to the owner list to be pinged to review future changes?

**Comment by jun-sheaf** [comment]
@andrewbranch I've just added myself to the owner list.

**Review by andrewbranch — APPROVED** [review]

**Comment by jun-sheaf** [comment]
Ready to merge

**Pull request merged**
